aten/src/THC/generic/THCTensorKthValue.cu

```diff
@@ -0,0 +1,122 @@
+  THCudaLongTensor_resize(state, indices, kthValueSize, NULL);
+  THLongStorage_free(kthValueSize);
+
+  #define RUN_K(INDEX_T, DIM)                                             \
+    gatherKthValue<real, INDEX_T, DIM>                                    \
+      <<<grid, block, 0, THCState_getCurrentStream(state)>>>(             \
+        inputInfo,                                                        \
+        sliceSize,                                                        \
+        k,                                                                \
+        inputSlices,                                                      \
+        /* The actual dimension that the k-selection is running in */     \
+        /* may have changed from collapseDims() */                        \
+        inputInfo.strides[collapseInputDim],                              \
+        kthValueInfo,                                                     \
+        indicesInfo)
+
+  #define RUN_DIM(INDEX_T)                        \
+    if (allDims == 1) {                           \
+      RUN_K(INDEX_T, 1);                          \
+    } else if (allDims == 2) {                    \
+      RUN_K(INDEX_T, 2);                          \
+    } else if (allDims == 3) {                    \
+      RUN_K(INDEX_T, 3);                          \
+    } else {                                      \
+      RUN_K(INDEX_T, -1);                         \
+    }
+
+  #define RUN_T(INDEX_T)                                                  \
+    TensorInfo<real, INDEX_T> inputInfo =                                 \
+      getTensorInfo<real, THCTensor, INDEX_T>(state, input);              \
+    TensorInfo<real, INDEX_T> kthValueInfo =                              \
+      getTensorInfo<real, THCTensor, INDEX_T>(state, kthValue);           \
+    TensorInfo<int64_t, INDEX_T> indicesInfo =                            \
+      getTensorInfo<int64_t, THCudaLongTensor, INDEX_T>(state, indices);  \
+                                                                          \
+    /* We use these structures solely to find the offset to */            \
+    /* each slice we are operating on */                                  \
+    inputInfo.sizes[dim] = 1;                                             \
+    kthValueInfo.sizes[dim] = 1;                                          \
+    indicesInfo.sizes[dim] = 1;                                           \
+                                                                          \
+    /* Collapse all other dims */                                         \
+    int collapseInputDim = inputInfo.collapseDims(dim);                   \
+    int collapseKthValueDim = kthValueInfo.collapseDims(dim);             \
+    int collapseIndicesDim = indicesInfo.collapseDims(dim);               \
+                                                                          \
+    int64_t inputSlices = 1;                                              \
```

Comment:
What exactly is a "inputSlice"?

Reply:
Suppose the input tensor shape is (M1, M2, M3, M4) and dim=2. We need to select the kth largest elements in the M3 dimension of every slice of the tensor, so there will be a total of M1xM2xM4 slices along which selection is to be done. We are parallelizing this across blocks by spawning a separate block for each slice along which we need to select, so number of blocks = inputSlices = M1xM2xM4.

This is the same logic as used in THCTensorTopK.cu, so I assumed it was self-explanatory. But I'm happy to add documentation to both files.

Reply:
Understood. Documentation would be good; I have a hard time reading THCTensorTopK.cu